src/matching/quoteMatcher.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * QuoteMatcher matches quoted strings, respecting escaped quotes (\") and frieds
+ */
+export class QuoteMatcher {
+  static escapeChar = "\\";
+
+  static findOpening(char: string, corpus: string, start: number): number {
+    // first, search backwards to see if we could be inside a quote
+    for (let i = start - 1; i >= 0; i--) {
+      const currentChar = corpus[i];
+      if (currentChar === char && corpus[i - 1] !== this.escapeChar) {
+        return i;
+      }
+    }
+
+    // Didn't find one behind us, the string may start ahead of us
+    for (let i = start; i < corpus.length; i++) {
+      const currentChar = corpus[i];
```

Comment:
I feel a bit bad (but not too bad) about copy-pasting the body of these loops. Really what I want is to have some thing that takes an iterator and returns Either<position **or** notfound> and then pass that iterator to the loops. But, I'm a total TypeScript noob and I didn't see an obvious way to do it.